### PR TITLE
Ty/pages new api

### PR DESCRIPTION
**Pull request by AlephNotation**
New api for browsers and pages. This is beginning of a major change for our API structure by exposing a lower level page and browser primitive

### CODE REVIEW

**Comment by matildepark** [comment]
Three fails, all based around timeouts. Still wondering what's going on there...

File: examples/wikipedia.ts [hunk 1]
```diff
@@ -40,10 +40,11 @@ async function main() {
     );
   }
   const logger = new Logger(["info"], (msg) => console.log(msg));
+  const agent = new Agent({ modelApi: chatApi });
 
   const agentBrowser = new AgentBrowser({
-    agent: new Agent({ modelApi: chatApi }),
-    browser: await Browser.create(argv.headless),
+    agent: agent,
```
Comment (matildepark) [inline examples/wikipedia.ts:46]: as a cute small thing, you can just do

```
agent,
```

as shorthand when the value is a var with the same name as the key

File: src/browser/browser.ts [hunk 2]
```diff
@@ -4,56 +4,120 @@ import { Page } from "./page";
...
+   *
+   * @param {boolean} headless - Specifies if the browser should be launched in headless mode.
+   * @param {Agent} agent - The agent that will interact with the browser.
+   * @param {Logger} [logger] - Optional logger to pass for browser operation logs.
+   * @param {string} [browserWSEndpoint] - The WebSocket endpoint to connect to a browser instance.
+   * @param {string[]} [browserLaunchArgs] - Additional arguments for launching the browser.
+   * @param {BrowserMode} [mode=BrowserMode.text] - The mode of the browser, defaults to text.
+   * @returns {Promise<Browser>} A promise that resolves to an instance of Browser.
+   */
+  static async launch(
```
Comment (matildepark) [inline src/browser/browser.ts:65]: Why are we renaming this? I am not sure what this accomplishes other than forcing everyone to change the method they use, when we seem to be keeping AgentBrowser so as to not do that.

Comment (AlephNotation) [reply src/browser/browser.ts:65]: trying to keep the puppeteer API as closely as possible https://github.com/puppeteer/puppeteer/blob/d8654f1eed657bb25da5d7167ed0abedf007d7a1/examples/pdf.js#L12

Comment (matildepark) [reply src/browser/browser.ts:65]: Okay. Well let me at least patch the `bin` folder here so our alpha package doesn't crash when you try to use it, I suppose.

**Comment by matildepark** [comment]
Sometimes we get a Zod error, which I've never seen before, when trying to run the thing.

`pnpm run build`, `node ./dist/bin/index.js --config ../config.json --objective "what's tynan's role?" --startUrl "https://hdr.is"` and...

```
⠹ Found Tynan Daly's role on the current pageZodError: [
  {
    "code": "invalid_union",
    "unionErrors": [
      {
        "issues": [
          {
            "received": "None",
            "code": "invalid_literal",
            "expected": "Type",
            "path": [
              "command",
              0,
              "kind"
            ],
            "message": "Invalid literal value, expected \"Type\""
          },
          {
            "code": "invalid_type",
            "expected": "number",
            "received": "undefined",
            "path": [
              "command",
              0,
              "index"
            ],
            "message": "Required"
          },
          {
            "code": "invalid_type",
            "expected": "string",
            "received": "undefined",
            "path": [
              "command",
              0,
              "text"
            ],
            "message": "Required"
          }
        ],
        "name": "ZodError"
      },
      {
        "issues": [
          {
            "received": "None",
            "code": "invalid_literal",
            "expected": "Click",
            "path": [
              "command",
              0,
              "kind"
            ],
            "message": "Invalid literal value, expected \"Click\""
          },
          {
            "code": "invalid_type",
            "expected": "number",
            "received": "undefined",
            "path": [
              "command",
              0,
              "index"
            ],
            "message": "Required"
          }
        ],
        "name": "ZodError"
      },
      {
        "issues": [
          {
            "received": "None",
            "code": "invalid_literal",
            "expected": "Wait",
            "path": [
              "command",
              0,
              "kind"
            ],
            "message": "Invalid literal value, expected \"Wait\""
          }
        ],
        "name": "ZodError"
      },
      {
        "issues": [
          {
            "received": "None",
            "code": "invalid_literal",
            "expected": "Back",
            "path": [
              "command",
              0,
              "kind"
            ],
            "message": "Invalid literal value, expected \"Back\""
          }
        ],
        "name": "ZodError"
      }
    ],
    "path": [
      "command",
      0
    ],
    "message": "Invalid input"
  }
]
    at get error (/Users/am/git/standard/hdr-browser/node_modules/.pnpm/zod@3.23.8/node_modules/zod/lib/types.js:55:31)
    at ZodObject.parse (/Users/am/git/standard/hdr-browser/node_modules/.pnpm/zod@3.23.8/node_modules/zod/lib/types.js:160:22)
    at chat (/Users/am/git/standard/hdr-browser/node_modules/.pnpm/zod-gpt@0.12.1_llm-api@1.5.1_zod@3.23.8/node_modules/zod-gpt/dist/src/completion.js:128:37)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async Agent._call (/Users/am/git/standard/hdr-browser/dist/agent/agent.js:119:26) {
  issues: [
    {
      code: 'invalid_union',
      unionErrors: [Array],
      path: [Array],
      message: 'Invalid input'
    }
  ],
  addIssue: [Function (anonymous)],
  addIssues: [Function (anonymous)],
  errors: [
    {
      code: 'invalid_union',
      unionErrors: [Array],
      path: [Array],
      message: 'Invalid input'
    }
  ]
}
✖ Agent failed to respond
```

**Comment by AlephNotation** [comment]
fixing

**Comment by AlephNotation** [comment]
I think i fixed it. I wasn't letting `command` be optional which seems to have done the fix. 
This should also fix the issue where the models find an answer and yet keep going
```
Debugger attached.
✔ Please specify an LLM provider for the agent · openai
✔ Please specify an LLM model for the agent · gpt-4
✔ Tynan Daly's role is in 'product, back-end'
Tynan Daly's role is in 'product, back-end'```

**Comment by matildepark** [comment]
Can corroborate. FTR we return the response twice there because the spinner doesn't return anything to what it's being piped to by itself, so we need to console log when we're done. I might just remove the spinner complete state so it doesn't do that, but that's separate, anyway.

**Comment by matildepark** [comment]
The test is failing on an invalid union, by the way.

```
ZodError: [
      {
        "code": "invalid_union",
        "unionErrors": [
          {
            "issues": [
              {
                "received": "type",
                "code": "invalid_literal",
                "expected": "Type",
                "path": [
                  "command",
                  0,
                  "kind"
                ],
                "message": "Invalid literal value, expected \"Type\""
              }
            ],
            "name": "ZodError"
          },
          {
            "issues": [
              {
                "received": "type",
                "code": "invalid_literal",
                "expected": "Click",
                "path": [
                  "command",
                  0,
                  "kind"
                ],
                "message": "Invalid literal value, expected \"Click\""
              }
            ],
            "name": "ZodError"
          },
          {
            "issues": [
              {
                "received": "type",
                "code": "invalid_literal",
                "expected": "Wait",
                "path": [
                  "command",
                  0,
                  "kind"
                ],
                "message": "Invalid literal value, expected \"Wait\""
              }
            ],
            "name": "ZodError"
          },
          {
            "issues": [
              {
                "received": "type",
                "code": "invalid_literal",
                "expected": "Back",
                "path": [
                  "command",
                  0,
                  "kind"
                ],
                "message": "Invalid literal value, expected \"Back\""
              }
            ],
            "name": "ZodError"
          },
          {
            "issues": [
              {
                "received": "type",
                "code": "invalid_literal",
                "expected": "Enter",
                "path": [
                  "command",
                  0,
                  "kind"
                ],
                "message": "Invalid literal value, expected \"Enter\""
              }
            ],
            "name": "ZodError"
          },
          {
            "issues": [
              {
                "received": "type",
                "code": "invalid_literal",
                "expected": "Hover",
                "path": [
                  "command",
                  0,
                  "kind"
                ],
                "message": "Invalid literal value, expected \"Hover\""
              }
            ],
            "name": "ZodError"
          },
          {
            "issues": [
              {
                "received": "type",
                "code": "invalid_literal",
                "expected": "Scroll",
                "path": [
                  "command",
                  0,
                  "kind"
                ],
                "message": "Invalid literal value, expected \"Scroll\""
              },
              {
                "expected": "'up' | 'down'",
                "received": "undefined",
                "code": "invalid_type",
                "path": [
                  "command",
                  0,
                  "direction"
                ],
                "message": "Required"
              }
            ],
            "name": "ZodError"
          }
        ],
        "path": [
          "command",
          0
        ],
        "message": "Invalid input"
```

`type` instead of `Type`, perhaps.

**Comment by AlephNotation** [comment]
Ah figured it out. the CMI was ingesting some bad memories and returning them without hints

**Comment by matildepark** [comment]
And should we worry about the invalid action generation we're getting sporadically? Those are, in my memory, kind of new for us. We don't usually choke on bad action generations, we usually just make it do it again, right?

**Pull request merged**